src/parquet/column_scanner.h

```diff
@@ -0,0 +1,149 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_COLUMN_SCANNER_H
+#define PARQUET_COLUMN_SCANNER_H
+
+#include <ostream>
+#include <string>
+#include <vector>
+
+namespace parquet_cpp {
+
+class ColumnReader;
+
+template <int TYPE>
+class TypedColumnReader;
+
+class Scanner {
+ public:
+  explicit Scanner(ColumnReader* reader) :
+      reader_(reader),
+      level_offset_(0),
+      levels_buffered_(0),
+      value_offset_(0),
+      values_buffered_(0) {
+    // TODO: don't allocate for required fields
+    def_levels_.resize(BATCHSIZE);
+    rep_levels_.resize(BATCHSIZE);
+  }
+
+  virtual void PrintNext(std::ostream& out, int width) = 0;
+
+  bool HasNext() {
+    return value_offset_ < values_buffered_ || reader_->HasNext();
+  }
+
+  bool NextLevels(int16_t* def_level, int16_t* rep_level) {
+    if (level_offset_ == levels_buffered_) {
+      levels_buffered_ = reader_->ReadDefinitionLevels(BATCHSIZE, &def_levels_[0]);
+
+      // TODO: repetition levels
+
+      level_offset_ = 0;
+      if (!levels_buffered_) {
+        return false;
+      }
+    }
+    *def_level = def_levels_[level_offset_++];
+    *rep_level = 1;
+    return true;
+  }
+
+ protected:
+  ColumnReader* reader_;
```

Comment:
If I read the code correctly, ColumnReader::GetScanner() returns a shared_ptr<Scanner>. So, it is possible that after a reader is destroyed, the scanner is still alive. To be usable (and avoid crashes), it needs reader_ .